File: AssetManager/UserInterface/Forms/Attachments/AttachmentObject.vb
Imports System.IO

Public Class SibiAttachment
    Inherits Attachment
    Public Property SelectedFolder As String

    Sub New(newFile As String, attachTable As AttachmentsBaseCols)
        MyBase.New(newFile, attachTable)
    End Sub

    Sub New(newFile As String, folderGUID As String, attachTable As AttachmentsBaseCols)
        MyBase.New(newFile, folderGUID, attachTable)
    End Sub

    Sub New(newFile As String, folderGUID As String, selectedFolder As String, attachTable As AttachmentsBaseCols)
        MyBase.New(newFile, folderGUID, attachTable)
        Me.SelectedFolder = selectedFolder
    End Sub

    Sub New(attachInfoTable As DataTable, selectedFolder As String, attachTable As AttachmentsBaseCols)
        MyBase.New(attachInfoTable, attachTable)
        Me.SelectedFolder = selectedFolder
    End Sub

End Class

Public Class DeviceAttachment
    Inherits Attachment

    Sub New(newFile As String, attachTable As AttachmentsBaseCols)
        MyBase.New(newFile, attachTable)
    End Sub

    Sub New(newFile As String, folderGUID As String, attachTable As AttachmentsBaseCols)
        MyBase.New(newFile, folderGUID, attachTable)
    End Sub

End Class

Public Class Attachment : Implements IDisposable
    Private _fileInfo As FileInfo
    Private _fileName As String
    Private _fileSize As Integer
    Private _extention As String
    Private _folderGUID As String
    Private _MD5 As String
    Private _fileUID As String
    Private _attachTable As AttachmentsBaseCols
    Private _dataStream As Stream

    Sub New()
        _fileInfo = Nothing
        _fileName = Nothing
        _fileSize = Nothing
        _extention = Nothing
        _folderGUID = Nothing
        _MD5 = Nothing
        _fileUID = Nothing
        _attachTable = Nothing
        _dataStream = Nothing
    End Sub

    ''' <summary>
    ''' Create new Attachment from a file path.
    ''' </summary>
    ''' <param name="NewFile">Full path to file.</param>
    Sub New(newFile As String, attachTable As AttachmentsBaseCols)
        _fileInfo = New FileInfo(newFile)
        _fileName = Path.GetFileNameWithoutExtension(_fileInfo.Name)
        _fileUID = Guid.NewGuid.ToString
        _MD5 = Nothing
        _fileSize = CInt(_fileInfo.Length)
        _extention = _fileInfo.Extension
        _folderGUID = String.Empty
        _attachTable = attachTable
        _dataStream = _fileInfo.OpenRead
    End Sub

    Sub New(newFile As String, folderGUID As String, attachTable As AttachmentsBaseCols)
        _fileInfo = New FileInfo(newFile)
        _fileName = Path.GetFileNameWithoutExtension(_fileInfo.Name)
        _fileUID = Guid.NewGuid.ToString
        _MD5 = Nothing
        _fileSize = CInt(_fileInfo.Length)
        _extention = _fileInfo.Extension
        _folderGUID = folderGUID
        _attachTable = attachTable
        _dataStream = _fileInfo.OpenRead
    End Sub

    Sub New(attachInfoTable As DataTable, attachTable As AttachmentsBaseCols)
        Dim TableRow As DataRow = attachInfoTable.Rows(0)
        _fileInfo = Nothing
        _dataStream = Nothing
        _attachTable = attachTable
        With TableRow
            _fileName = .Item(attachTable.FileName).ToString
            _fileUID = .Item(attachTable.FileUID).ToString
            _MD5 = .Item(attachTable.FileHash).ToString
            _fileSize = CInt(.Item(attachTable.FileSize))
            _extention = .Item(attachTable.FileType).ToString
            _folderGUID = .Item(attachTable.FKey).ToString
        End With
    End Sub

    Public ReadOnly Property FileInfo As FileInfo
        Get
            Return _fileInfo
        End Get
    End Property

    Public ReadOnly Property FileName As String
        Get
            If _fileInfo IsNot Nothing Then
                Return Path.GetFileNameWithoutExtension(_fileInfo.Name)
            Else
                Return _fileName
            End If
        End Get
    End Property

    Public ReadOnly Property FullFileName As String
        Get
            If _fileInfo IsNot Nothing Then
                Return _fileInfo.Name
            Else
                Return _fileName & Extension
            End If
        End Get
    End Property

    Public ReadOnly Property Extension As String
        Get
            If _fileInfo IsNot Nothing Then
                Return _fileInfo.Extension
            Else
                Return _extention
            End If

        End Get
    End Property

    Public ReadOnly Property Filesize As Long
        Get
            If _fileInfo IsNot Nothing Then
                Return _fileInfo.Length
            Else
                Return _fileSize
            End If
        End Get
    End Property

    Public ReadOnly Property FileUID As String
        Get
            Return _fileUID
        End Get
    End Property

    Public ReadOnly Property MD5 As String
        Get
            If _MD5 IsNot Nothing Then
                Return _MD5
            Else
                _MD5 = GetHash(_fileInfo)
                Return _MD5
            End If
        End Get
    End Property

    Public ReadOnly Property FolderGUID As String
        Get
            Return _folderGUID
        End Get
    End Property

    Public Property DataStream As Stream
        Get
            Return _dataStream
        End Get
        Set(value As Stream)
            _dataStream = value
        End Set
    End Property

    Public Property AttachTable As AttachmentsBaseCols
        Get
            Return _attachTable
        End Get
        Set(value As AttachmentsBaseCols)
            _attachTable = value
        End Set
    End Property

    Private Function GetHash(Fileinfo As FileInfo) As String
        Using HashStream As FileStream = Fileinfo.OpenRead
            Return SecurityTools.GetMD5OfStream(HashStream)
        End Using
    End Function

#Region "IDisposable Support"

    Private disposedValue As Boolean ' To detect redundant calls

    ' IDisposable
    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not disposedValue Then
            If disposing Then
                ' TODO: dispose managed state (managed objects).
                If _dataStream IsNot Nothing Then _dataStream.Dispose()
                _fileInfo = Nothing
            End If

            ' TODO: free unmanaged resources (unmanaged objects) and override Finalize() below.
            ' TODO: set large fields to null.
        End If
        disposedValue = True
    End Sub

    ' TODO: override Finalize() only if Dispose(disposing As Boolean) above has code to free unmanaged resources.
    'Protected Overrides Sub Finalize()
    '    ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
    '    Dispose(False)
    '    MyBase.Finalize()
    'End Sub

    ' This code added by Visual Basic to correctly implement the disposable pattern.
    Public Sub Dispose() Implements IDisposable.Dispose
        ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
        Dispose(True)
        ' TODO: uncomment the following line if Finalize() is overridden above.
        ' GC.SuppressFinalize(Me)
    End Sub

#End Region

End Class

File: AssetManager.ConvertedToC#/UserInterface/Forms/Attachments/AttachmentObject.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Threading.Tasks;
using System.IO;
namespace AssetManager
{

    public class Attachment : IDisposable
    {
        private FileInfo _fileInfo;
        private string _fileName;
        private int _fileSize;
        private string _extention;
        private string _folderName;
        private string _folderGUID;
        private string _MD5;
        private string _fileUID;
        private AttachmentsBaseCols _attachTable;

        private Stream _dataStream;
        public Attachment()
        {
            _fileInfo = null;
            _fileName = null;
            _fileSize = 0;
            _extention = null;
            _folderName = null;
            _folderGUID = null;
            _MD5 = null;
            _fileUID = null;
            _attachTable = null;
            _dataStream = null;
        }

        /// <summary>
        /// Create new Attachment from a file path.
        /// </summary>
        /// <param name="NewFile">Full path to file.</param>
        public Attachment(string newFile, AttachmentsBaseCols attachTable)
        {
            _fileInfo = new FileInfo(newFile);
            _fileName = Path.GetFileNameWithoutExtension(_fileInfo.Name);
            _fileUID = Guid.NewGuid().ToString();
            _MD5 = null;
            _fileSize = Convert.ToInt32(_fileInfo.Length);
            _extention = _fileInfo.Extension;
            _folderName = string.Empty;
            _folderGUID = string.Empty;
            _attachTable = attachTable;
            _dataStream = _fileInfo.OpenRead();
        }

        public Attachment(string newFile, string folderGUID, AttachmentsBaseCols attachTable)
        {
            _fileInfo = new FileInfo(newFile);
            _fileName = Path.GetFileNameWithoutExtension(_fileInfo.Name);
            _fileUID = Guid.NewGuid().ToString();
            _MD5 = null;
            _fileSize = Convert.ToInt32(_fileInfo.Length);
            _extention = _fileInfo.Extension;
            _folderName = string.Empty;
            _folderGUID = folderGUID;
            _attachTable = attachTable;
            _dataStream = _fileInfo.OpenRead();
        }

        public Attachment(DataTable attachInfoTable, AttachmentsBaseCols attachTable)
        {
            DataRow TableRow = attachInfoTable.Rows[0];
            _fileInfo = null;
            _dataStream = null;
            _attachTable = attachTable;
            
            _fileName = TableRow[attachTable.FileName].ToString();
            _fileUID = TableRow[attachTable.FileUID].ToString();
            _MD5 = TableRow[attachTable.FileHash].ToString();
            _fileSize = Convert.ToInt32(TableRow[attachTable.FileSize]);
            _extention = TableRow[attachTable.FileType].ToString();
            _folderName = TableRow[attachTable.Folder].ToString();
            _folderGUID = TableRow[attachTable.FKey].ToString();
        }

        public Attachment(string newFile, string folderGUID, string selectedFolder, AttachmentsBaseCols attachTable)
        {
            _fileInfo = new FileInfo(newFile);
            _fileName = Path.GetFileNameWithoutExtension(_fileInfo.Name);
            _fileUID = Guid.NewGuid().ToString();
            _MD5 = null;
            _fileSize = Convert.ToInt32(_fileInfo.Length);
            _extention = _fileInfo.Extension;
            _folderName = selectedFolder;
            _folderGUID = folderGUID;
            _attachTable = attachTable;
            _dataStream = _fileInfo.OpenRead();
        }

        public Attachment(DataTable attachInfoTable, string selectedFolder, AttachmentsBaseCols attachTable)
        {
            DataRow TableRow = attachInfoTable.Rows[0];
            _fileInfo = null;
            _dataStream = null;
            _attachTable = attachTable;
            var _with2 = TableRow;
            _fileName = TableRow[attachTable.FileName].ToString();
            _fileUID = TableRow[attachTable.FileUID].ToString();
            _MD5 = TableRow[attachTable.FileHash].ToString();
            _fileSize = Convert.ToInt32(TableRow[attachTable.FileSize]);
            _extention = TableRow[attachTable.FileType].ToString();
            _folderName = TableRow[attachTable.Folder].ToString();
            _folderGUID = TableRow[attachTable.FKey].ToString();
        }

        public FileInfo FileInfo
        {
            get { return _fileInfo; }
        }

        public string FileName
        {
            get
            {
                if (_fileInfo != null)
                {
                    return Path.GetFileNameWithoutExtension(_fileInfo.Name);
                }
                else
                {
                    return _fileName;
                }
            }
        }

        public string FullFileName
        {
            get
            {
                if (_fileInfo != null)
                {
                    return _fileInfo.Name;
                }
                else
                {
                    return _fileName + Extension;
                }
            }
        }

        public string Extension
        {
            get
            {
                if (_fileInfo != null)
                {
                    return _fileInfo.Extension;
                }
                else
                {
                    return _extention;
                }

            }
        }

        public long Filesize
        {
            get
            {
                if (_fileInfo != null)
                {
                    return _fileInfo.Length;
                }
                else
                {
                    return _fileSize;
                }
            }
        }

        public string FileUID
        {
            get { return _fileUID; }
        }

        public string MD5
        {
            get
            {
                if (_MD5 != null)
                {
                    return _MD5;
                }
                else
                {
                    _MD5 = GetHash(_fileInfo);
                    return _MD5;
                }
            }
        }

        public string FolderName
        {
            get { return _folderName; }
        }

        public string FolderGUID
        {
            get { return _folderGUID; }
        }

        public Stream DataStream
        {
            get { return _dataStream; }
            set { _dataStream = value; }
        }

        public AttachmentsBaseCols AttachTable
        {
            get { return _attachTable; }
            set { _attachTable = value; }
        }

        private string GetHash(FileInfo Fileinfo)
        {
            using (FileStream HashStream = Fileinfo.OpenRead())
            {
                return SecurityTools.GetMD5OfStream(HashStream);
            }
        }

        #region "IDisposable Support"

        // To detect redundant calls
        private bool disposedValue;

        // IDisposable
        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    // TODO: dispose managed state (managed objects).
                    if (_dataStream != null)
                        _dataStream.Dispose();
                    _fileInfo = null;
                }

                // TODO: free unmanaged resources (unmanaged objects) and override Finalize() below.
                // TODO: set large fields to null.
            }
            disposedValue = true;
        }

        // TODO: override Finalize() only if Dispose(disposing As Boolean) above has code to free unmanaged resources.
        //Protected Overrides Sub Finalize()
        //    ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
        //    Dispose(False)
        //    MyBase.Finalize()
        //End Sub

        // This code added by Visual Basic to correctly implement the disposable pattern.
        public void Dispose()
        {
            // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(true);
            // TODO: uncomment the following line if Finalize() is overridden above.
            // GC.SuppressFinalize(Me)
        }

        #endregion

    }
}
